src/pandas/array.h

```diff
@@ -20,9 +20,19 @@ class Array {
  public:
   virtual ~Array() {}
 
-  int64_t length() const { return length_; }
-  std::shared_ptr<DataType> type() const { return type_; }
-  DataType::TypeId type_id() const { return type_->type(); }
+  virtual int64_t length() const = 0;
+  // There are two methods to obtain the data type.
+  // The signature without a shared_ptr allows sub-classes
+  // to have a covariant return type, which eliminates the
+  // need/danger of doing a static_cast when dealing with
+  // a concrete sub-class. Ideally, the shared_ptr signature
+  // would suffice, but the compiler cannot treat a shared_ptr
+  // to a base class and a shared_ptr to a subclass as a
+  // covariant return type.
+  virtual TypePtr type() const = 0;
+  virtual const DataType& type_reference() const = 0;
```

Comment:
I don't have an intuitive feel for how these will end up getting used in practice. For example, you might have methods that operate on, in essence, `std::vector<DataType::TypeId>` obtained from all the columns of a table -- so having this operate on the against the base Array type without virtual functions calls would be advantageous. An example might be determining the "promoted type" for a table row (though you almost certain will want to compute this once and cache it, so virtual function calls are not that worrisome).